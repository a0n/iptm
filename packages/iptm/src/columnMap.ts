// #region impl
// tslint:disable:no-if-statement no-expression-statement no-object-mutation
// tslint:disable:readonly-array array-type no-delete no-let no-use-before-declare
type ColumnMapImpl<T> = {
  s: [Array<number>, Array<T>]
  children: { [key: string]: ColumnMapImpl<T> }
}
const stripInPlace = <T>(store: ColumnMapImpl<T>): ColumnMap<T> => {
  if (store.s[0].length === 0) {
    delete store.s
  }
  const children = Object.values(store.children)
  if (children.length === 0) {
    delete store.children
  } else {
    children.forEach(stripInPlace)
  }
  return store
}

const ColumnMapImpl = {
  /**
   * Creates a new empty column map.
   * We are using mutability, so everything needs to be fresh.
   */
  empty: <T>(): ColumnMapImpl<T> => ({
    s: [[], []],
    children: {},
  }),
  /**
   * Strips unused fields in place and returns the thing as a ColumnStore, never
   * to be mutated again
   */
  build: stripInPlace,
}

const lookup = <V>(m: { [k: string]: V }, k: string): V | undefined => m[k]

const getOrCreateInPlace = <T>(store: ColumnMapImpl<T>, key: string): ColumnMapImpl<T> => {
  const result = lookup(store.children, key)
  if (result !== undefined) {
    return result
  } else {
    const child = ColumnMapImpl.empty<T>()
    store.children[key] = child
    return child
  }
}

const isPrimitive = (key: any): boolean => {
  if (key === null) {
    return true
  }
  const type = typeof key
  if (type === 'function') {
    throw new Error('What do you think this is? unisonweb.org?')
  }
  if (type === 'object') {
    return false
  }
  return true
}

const updateInPlace = (obj: any, path: RA<string>, from: number, value: any): any => {
  if (from === path.length) {
    // at the end, just return the value and let the caller deal with storing it
    return value
  } else {
    const key = path[from]
    const child = obj[key]
    const childExists = child !== undefined
    const child1 = childExists ? child : {}
    const child2 = updateInPlace(child1, path, from + 1, value)
    // if the column store is canonical, I will never overwrite a scalar value,
    // and the from === path.length - 1 test is not necessary. But let's accept
    // non-canonical formats as well
    const mustUpdate = from === path.length - 1 || !childExists
    if (mustUpdate) {
      obj[key] = child2
    }
    return obj
  }
}

export const fromColumnMap = <T>(columns: ColumnMap<T>): RA<T> => {
  const rows: any = {}
  // first position is placeholder for the current index
  const path: Array<any> = [undefined]
  const addToRows = (store: ColumnMap<T>): void => {
    if (store.s !== undefined) {
      const [indices, values] = store.s
      if (values.length !== indices.length) {
        throw new Error()
      }
      for (let i = 0; i < values.length; i++) {
        const index = indices[i]
        const value = values[i]
        path[0] = index
        updateInPlace(rows, path, 0, value)
      }
    }
    if (store.children !== undefined) {
      const children = store.children
      Object.entries(children).forEach(([key, childStore]) => {
        path.push(key)
        addToRows(childStore)
        path.pop()
      })
    }
  }
  addToRows(columns)
  return Object.values(rows)
}

const addToValuesAndIndices = (store: ColumnMapImpl<any>, obj: any, index: number): void => {
  if (isPrimitive(obj)) {
    store.s[0].push(index)
    store.s[1].push(obj)
  } else {
    Object.entries(obj).forEach(([key, value]) => {
      const childStore = getOrCreateInPlace(store, key)
      addToValuesAndIndices(childStore, value, index)
    })
  }
}

export const toColumnMap = <T>(rows: RA<T>): ColumnMap<T> => {
  const rootStore: ColumnMapImpl<T> = ColumnMapImpl.empty()
  rows.forEach((row, index) => {
    addToValuesAndIndices(rootStore, row, index)
  })
  return ColumnMapImpl.build<T>(rootStore)
}

const maxIndex = (a: ColumnMap<any>, max: number): number => {
  let currentMax = max
  if (a.s) {
    const indices = a.s[0]
    if (indices.length > 0) {
      currentMax = Math.max(currentMax, indices[indices.length - 1])
    }
  }
  if (a.children) {
    Object.values(a.children).forEach(child => {
      currentMax = maxIndex(child, currentMax)
    })
  }
  return currentMax
}

const shiftIndices = <T>(a: ColumnMap<T>, offset: number): ColumnMap<T> => {
  const scalar: [RA<number>, RA<T>] | undefined = a.s
    ? [a.s[0].map(x => x + offset), a.s[1]]
    : undefined
  const children = a.children
    ? Object.entries(a.children).reduce(
        (acc, [k, v]) => {
          acc[k] = shiftIndices(v, offset)
          return acc
        },
        {} as { [key: string]: ColumnMap<T> },
      )
    : undefined
  return { s: scalar, children }
}

const concat0 = <T>(a: ColumnMap<T>, b: ColumnMap<T>): ColumnMap<T> => {
  const av = a.s || [[], []]
  const bv = b.s || [[], []]
  const ac = a.children || {}
  const bc = b.children || {}
  const i1 = av[0].concat(bv[0])
  const v1 = av[1].concat(bv[1])
  const scalar: typeof a.s = i1.length > 0 ? [i1, v1] : undefined
  const children: typeof a.children = { ...ac, ...bc }
  const keys = Object.keys(children)
  keys.forEach(key => {
    const childa: ColumnMap<T> = ac[key]
    const childb: ColumnMap<T> = bc[key]
    if (childa && childb) {
      children[key] = concat0(childa, childb)
    }
  })
  return { s: scalar, children: keys.length > 0 ? children : undefined }
}

// @ts-ignore
const concat = <T>(a: ColumnMap<T>, b: ColumnMap<T>): ColumnMap<T> => {
  const offset = maxIndex(a, -1) + 1
  const b1 = shiftIndices(b, offset)
  return concat0(a, b1)
}

type ColumnIteratorResult = {
  value: any
  hasValue: boolean
}

type ColumnIterator = {
  next: (index: number, r: ColumnIteratorResult) => void
}

const ColumnIterator = {
  of: (values: [RA<number>, RA<any>]): ColumnIterator => {
    let current = 0
    const [is, vs] = values
    return {
      next: (index: number, r: ColumnIteratorResult): void => {
        while (current < is.length && is[current] < index) {
          current++
        }
        const hasValue = is[current] === index
        r.hasValue = hasValue
        r.value = hasValue ? vs[current] : undefined
      },
    }
  },
}

type ColumnIteratorMap<T> = Readonly<{
  values?: ColumnIterator
  children?: { [key: string]: ColumnIteratorMap<any> }
}>

const ColumnIteratorMap = {
  of: <T>(m: ColumnMap<T>): ColumnIteratorMap<T> => {
    const values = m.s !== undefined ? ColumnIterator.of(m.s) : undefined
    if (m.children === undefined) {
      return { values }
    } else {
      const children: { [key: string]: ColumnIteratorMap<any> } = {}
      Object.entries(m.children).forEach(([key, value]) => {
        children[key] = ColumnIteratorMap.of(value)
      })
      return { values, children }
    }
  },
}

const iterate0 = <T>(im: ColumnIteratorMap<T>, index: number, rs: ColumnIteratorResult): void => {
  if (im.values) {
    im.values.next(index, rs)
    if (rs.hasValue) {
      return
    }
  }
  let result: any
  if (im.children) {
    Object.entries(im.children).forEach(([key, value]) => {
      iterate0(value, index, rs)
      if (rs.hasValue) {
        if (result === undefined) {
          result = {}
        }
        result[key] = rs.value
      }
    })
  }
  rs.hasValue = result !== undefined
  rs.value = result
}

const iterator = <T>(value: ColumnMap<T>): Iterator<T> => {
  const im = ColumnIteratorMap.of(value)
  let index: number = 0
  const rs: ColumnIteratorResult = {
    hasValue: false,
    value: undefined,
  }
  return {
    next: (): IteratorResult<T> => {
      iterate0(im, index, rs)
      if (rs.hasValue) {
        index += 1
        return { value: rs.value, done: false }
      } else {
        // from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols
        // value - any JavaScript value returned by the iterator. Can be omitted when done is true.
        return { done: true } as any
      }
    },
  }
}

const builder = <T>(): ColumnMapBuilder<T> => {
  const rootStore: ColumnMapImpl<T> = ColumnMapImpl.empty()
  let index = 0
  return {
    add: (value: T): void => addToValuesAndIndices(rootStore, value, index++),
    build: () => ColumnMapImpl.build<T>(rootStore),
  }
}

const iterable = <T>(value: ColumnMap<T>): Iterable<T> => ({
  [Symbol.iterator]: () => iterator(value),
})
// #endregion
export type RA<T> = ReadonlyArray<T>
export type ColumnMap<T> = Readonly<{
  s?: [RA<number>, RA<any>]
  children?: { [key: string]: ColumnMap<any> }
}>
export interface ColumnMapBuilder<T> {
  add: (value: T) => void
  build: () => ColumnMap<T>
}
export const ColumnMap = {
  of: toColumnMap,
  toArray: fromColumnMap,
  concat,
  iterable,
  iterator,
  builder,
}
