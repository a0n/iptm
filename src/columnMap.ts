// tslint:disable:no-if-statement no-object-mutation no-expression-statement no-shadowed-variable readonly-array
// tslint:disable:array-type no-delete no-let no-console
import * as shajs from 'sha.js'
import { toCbor } from './cbor'
import { CompressedArray, DagArray } from './dagArray'

export type RA<T> = ReadonlyArray<T>

type ColumnMapImpl<T> = {
  values: [Array<number>, Array<T>]
  children: { [key: string]: ColumnMapImpl<T> }
}
const stripInPlace = <T>(store: ColumnMapImpl<T>): ColumnMap<T> => {
  if (store.values[0].length === 0) {
    delete store.values
  }
  const children = Object.values(store.children)
  if (children.length === 0) {
    delete store.children
  } else {
    children.forEach(stripInPlace)
  }
  return store
}
const getCompressedSize = async (m: ColumnMap<any>): Promise<number> => {
  let result = 0
  if (m.values !== undefined) {
    // enable delta compression for indices
    result += await DagArray.getCompressedSize(m.values[0], { forceDelta: true })
    // let the compressor figure out if delta compression makes sense
    result += await DagArray.getCompressedSize(m.values[1])
  }
  if (m.children !== undefined) {
    for (const child of Object.values(m.children)) {
      result += await getCompressedSize(child)
    }
  }
  return result
}
const sha1 = (buffer: Buffer): string => {
  return shajs('sha1')
    .update(buffer.toString('hex'))
    .digest('hex')
}
const getBufferForSizing = (x: DagArray): Promise<Buffer> => {
  if (Array.isArray(x)) {
    return toCbor(x)
  } else {
    const ca = x as CompressedArray
    // not exactly accurate, since we would have to add the
    // compression type and reference in case of delta compression.
    // also, we assume that cbor dag will store a buffer without overhead
    // (no base64 or anything)
    return Promise.resolve(ca.d)
  }
}
const compressedSizeDedupImpl = async (
  m: ColumnMap<any>,
  sizes: { [key: string]: number },
): Promise<{ [key: string]: number }> => {
  if (m.values !== undefined) {
    // enable delta compression for indices
    const indices = await DagArray.compress(m.values[0], { forceDelta: true })
    const values = await DagArray.compress(m.values[1])
    const ib = await getBufferForSizing(indices)
    const vb = await getBufferForSizing(values)
    sizes[sha1(ib)] = ib.length
    sizes[sha1(vb)] = vb.length
  }
  if (m.children !== undefined) {
    for (const child of Object.values(m.children)) {
      await compressedSizeDedupImpl(child, sizes)
    }
  }
  return sizes
}
const getCompressedSizesDedup = (m: ColumnMap<any>): Promise<number> =>
  compressedSizeDedupImpl(m, {}).then(sizes => {
    console.log(sizes)
    return Object.values(sizes).reduce((x, y) => x + y, 0)
  })

const ColumnMapImpl = {
  /**
   * Creates a new empty column map.
   * We are using mutability, so everything needs to be fresh.
   */
  empty: <T>(): ColumnMapImpl<T> => ({
    values: [[], []],
    children: {},
  }),
  /**
   * Strips unused fields in place and returns the thing as a ColumnStore, never
   * to be mutated again
   */
  build: stripInPlace,
}

const getOrCreateInPlace = <T>(store: ColumnMapImpl<T>, key: string): ColumnMapImpl<T> => {
  const result = store.children[key]
  if (result !== undefined) {
    return result
  } else {
    const child = ColumnMapImpl.empty<T>()
    store.children[key] = child
    return child
  }
}

const isPrimitive = (key: any): boolean => {
  if (key === null) {
    return true
  }
  const type = typeof key
  if (type === 'function') {
    throw new Error('What do you think this is? unisonweb.org?')
  }
  if (type === 'object') {
    return false
  }
  return true
}

const updateInPlace = (obj: any, path: RA<string>, from: number, value: any): any => {
  if (from === path.length) {
    // at the end, just return the value and let the caller deal with storing it
    return value
  } else {
    const key = path[from]
    const child = obj[key]
    const childExists = child !== undefined
    const child1 = childExists ? child : {}
    const child2 = updateInPlace(child1, path, from + 1, value)
    // if the column store is canonical, I will never overwrite a scalar value,
    // and the from === path.length - 1 test is not necessary. But let's accept
    // non-canonical formats as well
    const mustUpdate = from === path.length - 1 || !childExists
    if (mustUpdate) {
      obj[key] = child2
    }
    return obj
  }
}

export const fromColumnMap = <T>(columns: ColumnMap<T>): RA<T> => {
  const rows: any = {}
  // first position is placeholder for the current index
  const path: Array<any> = [undefined]
  const addToRows = (store: ColumnMap<T>): void => {
    if (store.values !== undefined) {
      const [indices, values] = store.values
      if (values.length !== indices.length) {
        throw new Error()
      }
      for (let i = 0; i < values.length; i++) {
        const index = indices[i]
        const value = values[i]
        path[0] = index
        updateInPlace(rows, path, 0, value)
      }
    }
    if (store.children !== undefined) {
      const children = store.children
      Object.entries(children).forEach(([key, childStore]) => {
        path.push(key)
        addToRows(childStore)
        path.pop()
      })
    }
  }
  addToRows(columns)
  return Object.values(rows)
}

const addToValuesAndIndices = (store: ColumnMapImpl<any>, obj: any, index: number): void => {
  if (isPrimitive(obj)) {
    store.values[0].push(index)
    store.values[1].push(obj)
  } else {
    Object.entries(obj).forEach(([key, value]) => {
      const childStore = getOrCreateInPlace(store, key)
      addToValuesAndIndices(childStore, value, index)
    })
  }
}

export const toColumnMap = <T>(rows: RA<T>): ColumnMap<T> => {
  const rootStore: ColumnMapImpl<T> = ColumnMapImpl.empty()
  rows.forEach((row, index) => {
    addToValuesAndIndices(rootStore, row, index)
  })
  return ColumnMapImpl.build<T>(rootStore)
}

export interface ColumnMapBuilder<T> {
  add: (value: T) => void
  build: () => ColumnMap<T>
}

export const ColumnMapBuilder = {
  create: <T>(): ColumnMapBuilder<T> => {
    const rootStore: ColumnMapImpl<T> = ColumnMapImpl.empty()
    let index = 0
    return {
      add: (value: T): void => addToValuesAndIndices(rootStore, value, index++),
      build: () => ColumnMapImpl.build<T>(rootStore),
    }
  },
}

// const pick = <T>(a: ColumnMap<T>, indices: number[]): ColumnMap<T> => {
//   throw new Error()
// }

// @ts-ignore
const concat = <T>(a: ColumnMap<T>, b: ColumnMap<T>): ColumnMap<T> => {
  const offset = maxIndex(a, -1) + 1
  const b1 = shiftIndices(b, offset)
  return concat0(a, b1)
}

const concat0 = <T>(a: ColumnMap<T>, b: ColumnMap<T>): ColumnMap<T> => {
  const av = a.values || [[], []]
  const bv = b.values || [[], []]
  const ac = a.children || {}
  const bc = b.children || {}
  const i1 = av[0].concat(bv[0])
  const v1 = av[1].concat(bv[1])
  const values: typeof a.values = i1.length > 0 ? [i1, v1] : undefined
  const children: typeof a.children = { ...ac, ...bc }
  const keys = Object.keys(children)
  keys.forEach(key => {
    const childa: ColumnMap<T> = ac[key]
    const childb: ColumnMap<T> = bc[key]
    if (childa && childb) {
      children[key] = concat0(childa, childb)
    }
  })
  return { values, children: keys.length > 0 ? children : undefined }
}

const shiftIndices = <T>(a: ColumnMap<T>, offset: number): ColumnMap<T> => {
  const values: [RA<number>, RA<T>] | undefined = a.values
    ? [a.values[0].map(x => x + offset), a.values[1]]
    : undefined
  const children = a.children
    ? Object.entries(a.children).reduce(
        (acc, [k, v]) => {
          acc[k] = shiftIndices(v, offset)
          return acc
        },
        {} as { [key: string]: ColumnMap<T> },
      )
    : undefined
  return { values, children }
}

const maxIndex = (a: ColumnMap<any>, max: number): number => {
  let currentMax = max
  if (a.values) {
    const indices = a.values[0]
    if (indices.length > 0) {
      currentMax = Math.max(currentMax, indices[indices.length - 1])
    }
  }
  if (a.children) {
    Object.values(a.children).forEach(child => {
      currentMax = maxIndex(child, currentMax)
    })
  }
  return currentMax
}

export type TypedColumnMap<T> = Readonly<
  T extends object
    ? { children: { [K in keyof T]: TypedColumnMap<T[K]> } }
    : { values: [RA<number>, RA<T>] }
>
export type ColumnMap<T> = Readonly<{
  values?: [RA<number>, RA<any>]
  children?: { [key: string]: ColumnMap<any> }
}>
export const ColumnMap = {
  compressedSize: getCompressedSize,
  compressedSizeDedup: getCompressedSizesDedup,
  of: toColumnMap,
  toArray: fromColumnMap,
}
