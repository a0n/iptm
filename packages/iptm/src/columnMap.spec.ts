// tslint:disable:readonly-array no-expression-statement
import { ColumnMap, RA } from './columnMap'

const columnMapTest = (name: string, roundtrip: <T>(rows: RA<T>) => RA<T>) =>
  describe(name, () => {
    it('should work for normal time series data', () => {
      const rows = [{ x: 1, y: 2 }, { x: 2, y: 1 }]

      expect(ColumnMap.of(rows)).toMatchSnapshot()
      expect(roundtrip(rows)).toEqual(rows)
    })

    it('should properly deal with missing values', () => {
      const rows = [{ y: 2 }, { x: 2 }]

      expect(ColumnMap.of(rows)).toMatchSnapshot()
      expect(roundtrip(rows)).toEqual(rows)
    })

    it('should properly deal with nested objects', () => {
      const rows = [
        { type: 'start', data: { timestamp: 1 } },
        { type: 'stop', data: { reason: 'because' } },
      ]

      expect(ColumnMap.of(rows)).toMatchSnapshot()
      expect(roundtrip(rows)).toEqual(rows)
    })

    it.skip('should properly deal with nested arrays', () => {
      const rows = [['a', 'b', 'c'], ['a', 'c'], ['a']]

      expect(ColumnMap.of(rows)).toMatchSnapshot()
      // expect(roundtrip(rows)).toEqual(rows)
    })

    it('should properly deal with weird javascript crap', () => {
      const rows = [
        // unusual values as values
        {
          x: undefined,
          y: null,
          z: true,
          w: false,
          a: 1.0,
          b: NaN,
          c: Infinity,
          d: -Infinity,
        },
        // unusual values as keys
        { true: 1, false: 0, 1: 0, 0: 1, undefined: false, null: 1 },
      ]
      expect(ColumnMap.of(rows)).toMatchSnapshot()
      expect(roundtrip(rows)).toEqual(rows)
    })

    it('should not support function values', () => {
      const rows = [
        {
          x: (x: string) => x,
        },
      ]
      expect(() => ColumnMap.of(rows)).toThrow()
    })

    it('should not allow converting invalid column maps to array', () => {
      const cols = {
        o: {
          line: {
            s: [[0, 1], ['t800']],
          },
        },
      }
      expect(() => ColumnMap.toArray(cols)).toThrow()
    })

    it('should support concatenation', () => {
      const rows = [{ x: 1, y: 2 }, { x: 2, y: 1 }]
      const a = ColumnMap.of(rows)
      const b = ColumnMap.of(rows)
      expect(ColumnMap.toArray(ColumnMap.concat(a, b))).toEqual(rows.concat(rows))
    })
  })

const fromToRoundtrip = <T>(rows: RA<T>): RA<T> => ColumnMap.toArray(ColumnMap.of(rows))

const iterableRoundtrip = <T>(rows: RA<T>): RA<T> => [
  ...ColumnMap.iterable<T>(ColumnMap.of<T>(rows)),
]

const builderRoundtrip = <T>(rows: RA<T>): RA<T> => {
  const builder = ColumnMap.builder()
  rows.forEach(builder.add)
  const cm = builder.build()
  return ColumnMap.toArray(cm)
}

columnMapTest('toColumnMap/fromColumnMap', fromToRoundtrip)
// columnMapTest('toColumnMap/iterable', iterableRoundtrip)
// columnMapTest('builder/fromColumnMap', builderRoundtrip)
