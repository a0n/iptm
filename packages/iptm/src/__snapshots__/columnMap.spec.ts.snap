// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builder/fromColumnMap should properly deal with missing values 1`] = `
Object {
  "children": Object {
    "x": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          2,
        ],
      ],
    },
    "y": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          2,
        ],
      ],
    },
  },
}
`;

exports[`builder/fromColumnMap should properly deal with nested objects 1`] = `
Object {
  "children": Object {
    "data": Object {
      "children": Object {
        "reason": Object {
          "s": Array [
            Array [
              1,
            ],
            Array [
              "because",
            ],
          ],
        },
        "timestamp": Object {
          "s": Array [
            Array [
              0,
            ],
            Array [
              1,
            ],
          ],
        },
      },
    },
    "type": Object {
      "s": Array [
        Array [
          0,
          1,
        ],
        Array [
          "start",
          "stop",
        ],
      ],
    },
  },
}
`;

exports[`builder/fromColumnMap should properly deal with weird javascript crap 1`] = `
Object {
  "children": Object {
    "0": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          1,
        ],
      ],
    },
    "1": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          0,
        ],
      ],
    },
    "a": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          1,
        ],
      ],
    },
    "b": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          NaN,
        ],
      ],
    },
    "c": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          Infinity,
        ],
      ],
    },
    "d": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          -Infinity,
        ],
      ],
    },
    "false": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          0,
        ],
      ],
    },
    "null": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          1,
        ],
      ],
    },
    "true": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          1,
        ],
      ],
    },
    "undefined": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          false,
        ],
      ],
    },
    "w": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          false,
        ],
      ],
    },
    "x": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          undefined,
        ],
      ],
    },
    "y": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          null,
        ],
      ],
    },
    "z": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          true,
        ],
      ],
    },
  },
}
`;

exports[`builder/fromColumnMap should work for normal time series data 1`] = `
Object {
  "children": Object {
    "x": Object {
      "s": Array [
        Array [
          0,
          1,
        ],
        Array [
          1,
          2,
        ],
      ],
    },
    "y": Object {
      "s": Array [
        Array [
          0,
          1,
        ],
        Array [
          2,
          1,
        ],
      ],
    },
  },
}
`;

exports[`toColumnMap/fromColumnMap should properly deal with missing values 1`] = `
Object {
  "children": Object {
    "x": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          2,
        ],
      ],
    },
    "y": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          2,
        ],
      ],
    },
  },
}
`;

exports[`toColumnMap/fromColumnMap should properly deal with nested objects 1`] = `
Object {
  "children": Object {
    "data": Object {
      "children": Object {
        "reason": Object {
          "s": Array [
            Array [
              1,
            ],
            Array [
              "because",
            ],
          ],
        },
        "timestamp": Object {
          "s": Array [
            Array [
              0,
            ],
            Array [
              1,
            ],
          ],
        },
      },
    },
    "type": Object {
      "s": Array [
        Array [
          0,
          1,
        ],
        Array [
          "start",
          "stop",
        ],
      ],
    },
  },
}
`;

exports[`toColumnMap/fromColumnMap should properly deal with weird javascript crap 1`] = `
Object {
  "children": Object {
    "0": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          1,
        ],
      ],
    },
    "1": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          0,
        ],
      ],
    },
    "a": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          1,
        ],
      ],
    },
    "b": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          NaN,
        ],
      ],
    },
    "c": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          Infinity,
        ],
      ],
    },
    "d": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          -Infinity,
        ],
      ],
    },
    "false": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          0,
        ],
      ],
    },
    "null": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          1,
        ],
      ],
    },
    "true": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          1,
        ],
      ],
    },
    "undefined": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          false,
        ],
      ],
    },
    "w": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          false,
        ],
      ],
    },
    "x": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          undefined,
        ],
      ],
    },
    "y": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          null,
        ],
      ],
    },
    "z": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          true,
        ],
      ],
    },
  },
}
`;

exports[`toColumnMap/fromColumnMap should work for normal time series data 1`] = `
Object {
  "children": Object {
    "x": Object {
      "s": Array [
        Array [
          0,
          1,
        ],
        Array [
          1,
          2,
        ],
      ],
    },
    "y": Object {
      "s": Array [
        Array [
          0,
          1,
        ],
        Array [
          2,
          1,
        ],
      ],
    },
  },
}
`;

exports[`toColumnMap/iterable should properly deal with missing values 1`] = `
Object {
  "children": Object {
    "x": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          2,
        ],
      ],
    },
    "y": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          2,
        ],
      ],
    },
  },
}
`;

exports[`toColumnMap/iterable should properly deal with nested objects 1`] = `
Object {
  "children": Object {
    "data": Object {
      "children": Object {
        "reason": Object {
          "s": Array [
            Array [
              1,
            ],
            Array [
              "because",
            ],
          ],
        },
        "timestamp": Object {
          "s": Array [
            Array [
              0,
            ],
            Array [
              1,
            ],
          ],
        },
      },
    },
    "type": Object {
      "s": Array [
        Array [
          0,
          1,
        ],
        Array [
          "start",
          "stop",
        ],
      ],
    },
  },
}
`;

exports[`toColumnMap/iterable should properly deal with weird javascript crap 1`] = `
Object {
  "children": Object {
    "0": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          1,
        ],
      ],
    },
    "1": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          0,
        ],
      ],
    },
    "a": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          1,
        ],
      ],
    },
    "b": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          NaN,
        ],
      ],
    },
    "c": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          Infinity,
        ],
      ],
    },
    "d": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          -Infinity,
        ],
      ],
    },
    "false": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          0,
        ],
      ],
    },
    "null": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          1,
        ],
      ],
    },
    "true": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          1,
        ],
      ],
    },
    "undefined": Object {
      "s": Array [
        Array [
          1,
        ],
        Array [
          false,
        ],
      ],
    },
    "w": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          false,
        ],
      ],
    },
    "x": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          undefined,
        ],
      ],
    },
    "y": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          null,
        ],
      ],
    },
    "z": Object {
      "s": Array [
        Array [
          0,
        ],
        Array [
          true,
        ],
      ],
    },
  },
}
`;

exports[`toColumnMap/iterable should work for normal time series data 1`] = `
Object {
  "children": Object {
    "x": Object {
      "s": Array [
        Array [
          0,
          1,
        ],
        Array [
          1,
          2,
        ],
      ],
    },
    "y": Object {
      "s": Array [
        Array [
          0,
          1,
        ],
        Array [
          2,
          1,
        ],
      ],
    },
  },
}
`;
